### PR TITLE
Add CodeAction & CodeActionKind support.

### PR DESCRIPTION
**Pull request by yaohaizh**
Signed-off-by: Yaohai Zheng <yaozheng@microsoft.com>

### CODE REVIEW

File: org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend [hunk 1]
```diff
@@ -546,17 +546,66 @@ class ImplementationCapabilities extends DynamicRegistrationCapabilities {
...
+
+@JsonRpcData
+class CodeActionKindCapabilities {
+	/**
+	 * The code action kind values the client supports. When this
+	 * property exists the client also guarantees that it will
+	 * handle values outside its set gracefully and falls back
+	 * to a default value when unknown.
+	 */
+    List<CodeActionKind> valueSet
```
Comment (spoenemann) [inline org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend]: This should be `List<String>`.

File: org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend [hunk 2]
```diff
@@ -778,6 +827,55 @@ class ClientCapabilities {
...
+     */
+    @NonNull
+	String title
+
+    /**
+     * The kind of the code action.
+     *
+     * Used to filter code actions.
+     */
+    CodeActionKind kind
```
Comment (spoenemann) [inline org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend]: This should be `String`.

File: org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend [hunk 3]
```diff
@@ -788,14 +886,26 @@ class CodeActionContext {
...
 	List<Diagnostic> diagnostics
-    
+	
+	/**
+	 * Requested kind of actions to return.
+	 *
+	 * Actions not of this kind are filtered out by the client before being shown. So servers
+	 * can omit computing them.
+	 */
+	List<CodeActionKind> only
```
Comment (spoenemann) [inline org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/Protocol.xtend]: This should be `List<String>`.

**Comment by yaohaizh** [comment]
@spoenemann Thanks. Updated the PR.

File: org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/services/TextDocumentService.java [hunk 4]
```diff
@@ -159,7 +160,7 @@
 	 * Registration Options: TextDocumentRegistrationOptions
 	 */
 	@JsonRequest
-	CompletableFuture<List<? extends Command>> codeAction(CodeActionParams params);
+	CompletableFuture<Either<List<Command>, List<CodeAction>>> codeAction(CodeActionParams params);
```
[the remark this answers is not in the document]
Comment (yaohaizh) [reply org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/services/TextDocumentService.java]: @spoenemann Fixed this

**Pull request merged**